# Machine learning-Iris classification
Welcome to the first day of <font color = "blue">#60daysofcode</font> proposed by Udacity.I decided to put what I learnt from this Nanodegree in practice so I started by lesson 3 and exactly **classification** problem so today I'm  going to develop the model that is going to **classify** the **iris** flowers for us.

The data used here is a famous dataset that contains the sepal and petal length and width of 150 iris flowers of three different species: Iris-Setosa, Iris-Versicolor, and Iris-Virginica.

Let’s try to create the model that can classify the different species of the Iris flower.

**Problem solving:**

*   create the dataset.
*   Exploring the data
*   Build the model
*   Train the model
*   Make predictions

**1.Create the datasets:**
In order to classify the different species of the Iris,We should prepare the datasets with **features** and **labels**.But sklearn comes with the inbuilt datasets for the iris classification problem, so no need to worry :) .

Let us first understand the datasets

The data set consists of:

*   150 samples
*   3 labels: species of Iris (Iris setosa, Iris virginica and Iris versicolor)
*   4 features: Sepal length,Sepal width,Petal length,Petal Width in cm

You need to note that Scikit learn only works if data is stored as numeric data, irrespective of it being a **regression** or a **classification** problem. It also requires the arrays to be stored at **numpy arrays** for optimization. 

Since, this dataset is loaded from scikit learn, everything is appropriately formatted.

Let's jump to the first line of code and undrestand it all :) :) 

First of all we need to load the Iris dataset for that we will import our data from sklearn library and we will need to use numpy for after.

Second step is to make a variable named "iris" (just to make it simple for us to undrestand but you can call it by your name don't worry :) ) and use <font color="blue"> datasets.load_iris() </font> that's load and return the iris dataset (classification).

In [0]:
from  sklearn import  datasets
import numpy as np
iris=datasets.load_iris()

**Machine learning terminology:**
* Each row is an **observation** (also known as: sample, example, instance, record)
* Each column is a **feature** (also known as: predictor, attribute, independent variable, input, regressor, covariate)

**2.Exploring the iris dataset:**


In [3]:
# print the names of the four features
print (iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


* Each value we are predicting is the response (also known as: target, outcome, label, dependent variable)
* **Classification** is supervised learning in which the response is categorical
  * "0": setosa
  * "1": versicolor
  * "2": virginica

In [4]:
# print integers representing the species of each observation
# 0, 1, and 2 represent different species
print (iris.target)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Assign the data and target to separate variables.
* x contains the features.
* y contains the labels.

In [0]:
x=iris.data
y=iris.target

In [6]:
list(iris.keys())

['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename']

**Splitting the dataset:**  
Since our process involve **training** and **testing** ,We should split our dataset.It can be executed by the following code : 

In [0]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.5)

The first line: from sklearn.model_selection import train_test_split (what? what That mean??) that's mean From Sklearn, sub-library model_selection, I’ve imported the train_test_split so I can, well, split to training and test sets   
**Note that :**
* <font color ="blue">x_train</font> contains the <font color ="blue">training features</font>
* <font color ="green">x_test</font> contains the <font color ="green">testing features</font>
* <font color ="blue">y_train</font> contains the <font color ="blue">training label</font>
* <font color ="green">y_test</font> contains the <font color ="green">testing labels</font>

And you use train_test_split(x,y,test_size=.5) to indicate:  
* **x:**  iris.data (seen above)  
* **y:**  iris.target (seen above)  
* **test_size:** float, int or None, optional (default=None)  
If float, should be between 0.0 and 1.0 and represent the proportion of the dataset to include in the test split. If int, represents the absolute number of test samples. If None, the value is set to the complement of the train size. If train_size is also None, it will be set to 0.25.

You didn't undrestand this part well? don't worry you can have a look at this exemple and try to implement it [Understanding the data splitting functions in scikit-learn](https://medium.com/@julie.yin/understanding-the-data-splitting-functions-in-scikit-learn-9ae4046fbd26)


**3.Build the model**  
We can use any classification algorithm to solve the problem.In this exemple I will use Logistic Regression because I saw it in the course ;) 

This code will create the empty model. In order to provide the operations to the model we should train them.  
**(Should I use just this model?)** Of course no you can use the model you want and even try different ones to compare your accuracy by the end.

In [0]:
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(random_state=0)

**You can't udrestand the code above?**  
 you can have a look at the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html) or this tutorial [Building A Logistic Regression in Python, Step by Step](https://towardsdatascience.com/building-a-logistic-regression-in-python-step-by-step-becd4d56c9c8)

**4.Train the Model:**  
We can train the model with fit function. (I will explain it to you :) )



In [9]:
classifier.fit(x_train,y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=0, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

As you can see, we’re fitting the model on the training data and trying to predict the test data.   

Now the model is ready to make predictions


**4.Make predictions:**  
Now that we have trained our algorithm, it’s time to make some predictions. To do so, we will use our test data and predict function and see how accurately our algorithm do. To make predictions on the test data, execute the following script:

In [21]:
predictions=classifier.predict(x_test)
predictions

array([1, 2, 2, 2, 0, 0, 1, 2, 0, 2, 0, 2, 2, 2, 0, 1, 0, 0, 2, 1, 2, 2,
       1, 2, 1, 1, 1, 0, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 2, 1, 0,
       2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 0, 2, 2, 0, 0, 2, 0, 2, 2, 0, 0, 2,
       1, 1, 2, 0, 0, 1, 1, 1, 2])

these predictions can be matched with the expected output to measure the accuracy value.

In [12]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,predictions))

0.9733333333333334


So the accuracy is 97% which is no bad :) 

This is it for this first day, that took me around 2 hours to make this document but I feel I undrestand more classification and Logistic Regression.
Looking forward to practice tomorrow :).